### PR TITLE
Wrap the input tables

### LINKED ISSUE #3034: Wrap the InputTable API in the DH Python API (server-side)
This is a feature most requested by the DevRel team.

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes #3034

### CODE REVIEW

File: py/server/deephaven/table_factory.py [hunk 1]
```diff
@@ -3,13 +3,14 @@
 #
 
 """ This module provides various ways to make a Deephaven table. """
-from typing import List, Dict, Any, Union
+from __future__ import annotations
```
Comment (devinrsmith) [inline py/server/deephaven/table_factory.py:6]: What's this import used for?

File: py/server/tests/test_table_factory.py [hunk 2]
```diff
@@ -200,11 +200,76 @@ def test_dtw_single_string_arg(self):
...
+            s_array = dtypes.array(dtypes.short, [128, 228, 328])
+            i_array = dtypes.array(dtypes.int32, [32768, 42768, 52768])
+            l_array = dtypes.array(dtypes.long, [2 ** 32, 2 ** 33, 2 ** 36])
+            f_array = dtypes.array(dtypes.float32, [1.0, 1.1, 1.2])
+            d_array = dtypes.array(dtypes.double, [1.0 / 2 ** 32, 1.1 / 2 ** 33, 1.2 / 2 ** 36])
+            str_array = dtypes.array(dtypes.string, ["some", "not so random", "text"])
+            table_writer.write_row(b_array, s_array, i_array, l_array, f_array, d_array, str_array
+                                   )
+        t = table_writer.table
+        self.wait_ticking_table_update(t, row_count=1, timeout=5)
```
Comment (devinrsmith) [inline py/server/tests/test_table_factory.py:229]: For clarity, I think this table that we are using as input to the inputs tables should be created via new_table instead of table writer.

File: py/server/tests/test_table_factory.py [hunk 3]
```diff
@@ -200,11 +200,76 @@ def test_dtw_single_string_arg(self):
...
+    def test_input_table(self):
+        col_defs = {
+            "ByteArray": dtypes.byte_array,
+            "ShortArray": dtypes.short_array,
+            "Int32Array": dtypes.int32_array,
+            "LongArray": dtypes.long_array,
+            "Float32Array": dtypes.float32_array,
+            "DoubleArray": dtypes.double_array,
+            "StringArray": dtypes.string_array,
+        }
```
Comment (devinrsmith) [inline py/server/tests/test_table_factory.py:217]: Why are these tests oriented to array types instead of primitives? Nothing "wrong" with that, I just would have expected something around primitive types.

File: py/server/tests/test_table_factory.py [hunk 4]
```diff
@@ -200,11 +200,76 @@ def test_dtw_single_string_arg(self):
...
+            append_only_input_table = InputTable(init_table=t)
+            with self.assertRaises(DHError) as cm:
+                append_only_input_table.delete(t)
+            self.assertIn("not allowed.", str(cm.exception))
+
+            keyed_input_table = InputTable(init_table=t, key_cols="StringArray")
+            self.assertEqual(keyed_input_table.size, 2)
+            keyed_input_table.delete(t.select("StringArray"))
+            self.assertEqual(keyed_input_table.size, 0)
+
```
Comment (devinrsmith) [inline py/server/tests/test_table_factory.py:272]: We should probably have a test w/ key_cols as array.

File: py/server/deephaven/table_factory.py (hunk 1 above)
Comment (jmao-denver) [reply py/server/deephaven/table_factory.py:6]: it was needed for an earlier iteration, no longer needed but auto-import-org didn't remove it for some reason.

File: py/server/tests/test_table_factory.py [hunk 5]
```diff
@@ -206,56 +206,42 @@ def test_dtw_single_string_arg(self):
...
-            str_array = dtypes.array(dtypes.string, ["some", "not so random", "text"])
-            table_writer.write_row(b_array, s_array, i_array, l_array, f_array, d_array, str_array
-                                   )
-        t = table_writer.table
-        self.wait_ticking_table_update(t, row_count=1, timeout=5)
-
+        cols = [
+            bool_col(name="Boolean", data=[True, False]),
+            byte_col(name="Byte", data=(1, -1)),
+            char_col(name="Char", data='-1'),
```
Comment (devinrsmith) [inline py/server/tests/test_table_factory.py:212]: How does this work? Shouldn't it be two items?

Comment (devinrsmith) [reply py/server/tests/test_table_factory.py:212]: Got it - so there are two characters here `-` and `1`. Thanks.

**Review by devinrsmith — APPROVED** [review]

File: py/server/deephaven/table_factory.py [hunk 6]
```diff
@@ -186,3 +194,77 @@ def write_row(self, *values: Any) -> None:
             self._j_table_writer.logRowPermissive(values)
         except Exception as e:
             raise DHError(e, "failed to write a row.") from e
+
+
+class InputTable(Table):
+    """InputTable is a subclass of Table that allows the users to dynamic add/delete/modify data in it. There are two
```
Comment (chipkent) [inline py/server/deephaven/table_factory.py:200]:
```suggestion
    """InputTable is a subclass of Table that allows the users to dynamically add/delete/modify data in it. There are two
```

File: py/server/deephaven/table_factory.py [hunk 7]
```diff
@@ -186,3 +194,77 @@ def write_row(self, *values: Any) -> None:
...
+        """
+        try:
+            if col_defs is None and init_table is None:
+                raise ValueError("either column definitions or init table should be provided.")
+            elif col_defs and init_table:
+                raise ValueError("both column definitions and init table are provided.")
+
+            if col_defs:
+                cols = [Column(name=n, data_type=t) for n, t in col_defs.items()]
+                j_arg_1 = _JTableDefinition.of([col.j_column_definition for col in cols])
```
Comment (chipkent) [inline py/server/deephaven/table_factory.py:226]: These could be a 1-liner

File: py/server/deephaven/table_factory.py [hunk 8]
```diff
@@ -186,3 +194,77 @@ def write_row(self, *values: Any) -> None:
...
+            if key_cols:
+                super().__init__(_JKeyedArrayBackedMutableTable.make(j_arg_1, key_cols))
+            else:
+                super().__init__(_JAppendOnlyArrayBackedMutableTable.make(j_arg_1))
+            self.j_input_table = self.j_table.getAttribute(_J_INPUT_TABLE_ATTRIBUTE)
+            self.key_columns = key_cols
+        except Exception as e:
+            raise DHError(e, "failed to create a InputTable.") from e
+
+    def add(self, table: Table):
```
Comment (chipkent) [inline py/server/deephaven/table_factory.py:240]: Are we annotating the `None` return type?  I don't remember what standard we are doing or what is best for the IDE.  Maybe what is there is best.

File: py/server/deephaven/table_factory.py [hunk 9]
```diff
@@ -186,3 +194,77 @@ def write_row(self, *values: Any) -> None:
...
+
+        Raises:
+            DHError
+        """
+        try:
+            self.j_input_table.add(table.j_table)
+        except Exception as e:
+            raise DHError(e, "add to InputTable failed.") from e
+
+    def delete(self, table: Table):
```
Comment (chipkent) [inline py/server/deephaven/table_factory.py:255]: Are we annotating the `None` return type?  I don't remember what standard we are doing or what is best for the IDE.  Maybe what is there is best.

File: py/server/deephaven/table_factory.py [hunk 10]
```diff
@@ -186,3 +194,77 @@ def write_row(self, *values: Any) -> None:
...
+            else:
+                super().__init__(_JAppendOnlyArrayBackedMutableTable.make(j_arg_1))
+            self.j_input_table = self.j_table.getAttribute(_J_INPUT_TABLE_ATTRIBUTE)
+            self.key_columns = key_cols
+        except Exception as e:
+            raise DHError(e, "failed to create a InputTable.") from e
+
+    def add(self, table: Table):
+        """Writes rows in the provided table to this input table. If this is keyed input table, added rows with keys
+        that match existing rows will replace those rows.
```
Comment (chipkent) [inline py/server/deephaven/table_factory.py:242]:
```suggestion
        """Writes rows from the provided table to this input table. If this is a keyed input table, added rows with keys
        that match existing rows will replace those rows.
```

Comment (chipkent) [reply py/server/deephaven/table_factory.py:242]: What happens if keyed-rows are duplicated?  Error?  Only the last one is used?

File: py/server/deephaven/table_factory.py [hunk 11]
```diff
@@ -186,3 +194,77 @@ def write_row(self, *values: Any) -> None:
...
+        Raises:
+            DHError
+        """
+        try:
+            self.j_input_table.add(table.j_table)
+        except Exception as e:
+            raise DHError(e, "add to InputTable failed.") from e
+
+    def delete(self, table: Table):
+        """Deletes the keys contained in the provided table from this keyed input table.
```
Comment (chipkent) [inline py/server/deephaven/table_factory.py:256]: Doc string is written as if the input table is always a keyed input table, but it could be keyed or append.

**Review by chipkent — CHANGES_REQUESTED** [review]

**Review by devinrsmith — DISMISSED** [review]

File: py/server/deephaven/table_factory.py (hunk 10 above)
Comment (chipkent) [reply py/server/deephaven/table_factory.py]: What happens if you have a keyed table and the added table contains duplicates of keyed rows?  Does that result in an error?  Is there no error, but the last value is used?  etc.

Comment (jmao-denver) [reply py/server/deephaven/table_factory.py]: From reading the Java implementation code and adding a test case, the answer to the questions is: no error will be used, and yes the last duplicate value is used.

**Review by chipkent — APPROVED** [review]

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

How-to: https://github.com/deephaven/deephaven.io/issues/1892
Conceptual: https://github.com/deephaven/deephaven.io/issues/1894
Reference: https://github.com/deephaven/deephaven.io/issues/1893
Blog: https://github.com/deephaven/deephaven.io/issues/1891
